### PR TITLE
VirtualizedList CellRenderer optimization

### LINKED ISSUE #20174: FlatList component inefficient function passing for CellRenderer
- [x] Review the documentation
  - [x] Search for existing issues
  - [ ] Use the latest React Native release - *using version 0.55.2 did not see any changes to the FlatListComponent since that version.*

## Environment
Run `react-native info` in your terminal and paste its contents here.
```
Environment:
  OS: Windows 10
  Node: 8.11.1
  Yarn: 0.21.3
  npm: 5.8.0
  Watchman: Not Found
  Xcode: N/A
  Android Studio: Version  3.0.0.0 AI-171.4408382

Packages: (wanted => installed)
  react: 16.3.1 => 16.3.1
  react-native: ^0.55.2 => 0.55.4

```
## Description
I used the npm package why-did-you-update to help optimize my components and prevent re-renders. However there is a re-render that I cannot control. This happens for every item in the list.  My gues

```
CellRenderer.props: Changes are in functions only. Possibly avoidable re-render?
Functions before: {onLayout: ƒ}
Functions after: {onLayout: ƒ}
```

It seems this is caused by an anonymous function at https://github.com/facebook/react-native/blob/d756d94b3a7e2812f17f549c57767ac63734b49c/Libraries/Lists/VirtualizedList.js#L691

I propose that this function be defined on the class to prevent unnecessary re-renders.

### PR DESCRIPTION
**Pull request by christianbach**
# Fixes

Fixes #20174 based of @jasekiw pull request #20208

# Changes
* Remove parentProps from CellRenderer and pass as ordinary props
* CellRenderer is now a PureComponent

Test Plan:
----------
Tested on a real device with why-did-you-update running to view when the CellRendere component re-rendered. Applied updates to the list items to make sure they still get updated.

Added a test that makes sure the initial onLayout is still called properly.

Release Notes:
--------------
[GENERAL] [ENHANCEMENT][VirtualList] - Prevent unnecessary re-renders of the CellRenderer Component.

### CODE REVIEW

**Comment by hramos** [comment]
Thanks for adding tests and @mattharget for approving this. I'll import this and put it on the team's review queue, as an additional precaution.

**Comment by sahrens** [comment]
Sorry for the delay - thanks for making this change.

Did you test to make sure that sticky headers in RNTester still work?

**Comment by christianbach** [comment]
No unfortunately I didn’t. I’ll check and report the results. But I checked all the list files for additional props and didn’t find any other references. Does the sticky headers implemention need CellRenderer to re-render?

**Comment by hramos** [comment]
When this patch is applied, I get a RedScreen on RNTester, specifically on VirtualizedList, VirtualizedSectionList.js:201: `Invariant Violation: Tried to get frame for out of range index NaN`

**Comment by christianbach** [comment]
Thanks @hramos I just booted up RNTester my self and got the same error. My bad for only testning this on FlatList and SectionLists without Sticky Headers.

**Comment by christianbach** [comment]
@sahrens I'm I correct that the problems lays in the fact that `onLayout` is extracted by ScrollViewStickyHeader and since it doesn't get the props from cellRenderer we receive a no cellKey or index?

**Comment by christianbach** [comment]
@sahrens @hramos Updated the PR and now RNTester/SectionList works as expected. Thanks for testing!

**Comment by christianbach** [comment]
@sahrens @hramos Any updates? I know your super busy, but a update would be appreciated :)

**Comment by christianbach** [comment]
@hramos can I do anything to help speed up the review?

**Comment by hramos** [comment]
@christianbach thanks for adding the tests, I'm merging this.

**Comment by christianbach** [comment]
@hramos any updates? 😸

File: Libraries/Components/ScrollView/ScrollViewStickyHeader.js [hunk 1]
```diff
@@ -58,8 +58,8 @@ class ScrollViewStickyHeader extends React.Component<Props, State> {
 
     this.props.onLayout(event);
     const child = React.Children.only(this.props.children);
-    if (child.props.onLayout) {
-      child.props.onLayout(event);
+    if (child.props._onLayout) {
+      child.props._onLayout(event);
```
[the remark this answers is not in the document]
Comment (christianbach) [reply Libraries/Components/ScrollView/ScrollViewStickyHeader.js]: Hi @sebryu and thanks for reviewing this!

`_onLayout` is available and is part of the rewrite of CellRenderer to stop functions being created on every render

```
_onLayout = (e): void =>
     this.props.onLayout &&
     this.props.onLayout(e, this.props.cellKey, this.props.index);
```

[the remark this answers is not in the document]
Comment (sebryu) [reply Libraries/Components/ScrollView/ScrollViewStickyHeader.js]: Hello @christianbach !
I read your code, and I noticed _onLayout function in CellRenderer class. However - you are trying to call it as `child.props._onLayout`, and simply - `_onLayout` function is not accessible from `child.props`. You would need to have ref to CellRenderer to call its functions.

So in this case, my suggested changes should work.

[the remark this answers is not in the document]
Comment (christianbach) [reply Libraries/Components/ScrollView/ScrollViewStickyHeader.js]: My bad and thanks! I will change it straight away!

[the remark this answers is not in the document]
Comment (christianbach) [reply Libraries/Components/ScrollView/ScrollViewStickyHeader.js]: Finally had the time to do a commit. Tested the change in our production app without issues.

[the remark this answers is not in the document]
Comment (christianbach) [reply Libraries/Components/ScrollView/ScrollViewStickyHeader.js]: @sebryu Do I need to do anything else, perhaps rebase from master?

[the remark this answers is not in the document]
Comment (sebryu) [reply Libraries/Components/ScrollView/ScrollViewStickyHeader.js]: Hi @christianbach !

You have to ping RN guys, I just wanted to create similar issue, so I reviewed your code. 
It looks ok for me though 👍

**Comment by stanislavgalin** [comment]
Any updates, guys @hramos @sahrens @matthargett ?

**Comment by hramos** [comment]
Hey folks, please avoid pinging people unnecessarily. We'll get through the queue as fast as we can, and provide updates in this thread as they become available.

**Comment by christianbach** [comment]
@hramos i would love to know what failed and how I can move this along.

**Comment by hramos** [comment]
@christianbach The problem was unrelated to your code changes, the tool had trouble applying your changes given we took too long to import them and the parent commit was way behind. I copied your changes and rebased them on top of the latest master, and they should be landing any minute now. We should see a message from @facebook-github-bot once that happens.

**Comment by hramos** [comment]
Unfortunately, this introduces a failure in one of our internal customer's tests. It's going to be a bit awkward to work through this, as the test case is not visible on GitHub. We'll likely need to resolve this on our end at some point. I'm just sharing here for transparency.

**Comment by christianbach** [comment]
Thanks @hramos for taking the time trying to get this merged.

**Comment by noomorph** [comment]
@christianbach, I had an opportunity to give your pull request a test drive, and I found one error.

You have removed passing parent props but as a consequence, list stopped rerendering if it depended on extraData={}.

https://github.com/facebook/react-native/blob/1151c096dab17e5d9a6ac05b61aacecd4305f3db/Libraries/Lists/VirtualizedList.js#L701

I suggest to fix it in your PR - you ought to pass extraData there. What do you think?

**Comment by christianbach** [comment]
@noomorph sounds reasonable to me. Your using extraData to force a rerendering of the whole flatlist right? So updating extraData will not be shallow equal to the old extra data or undefined and that will force the Pure CellRenderer to update. 

I’m AFK until Monday but after that I can take a look.

**Comment by noomorph** [comment]
Yes, I can give an example:

```javascript
        <FlatList
          data={filteredItems}
          extraData={selectedItemsIds}
          keyExtractor={this.getItemKey}
          getItemLayout={this.getItemLayout}
          renderItem={this.renderItem}
        />
```

As you see in this case that was broken - the items were separated from their respective selected statues, which is an arguable solution but still technically valid, and it skips renders when only `selectedItemsIds` is being changed between parent component rerenders.

**Comment by christianbach** [comment]
@noomorph I updated the code to now support the extraData prop, I added a test as well so hopefully everything will work for you as intended.

**Comment by noomorph** [comment]
@christianbach, amazing! Thanks much, I'll test your changes during this week.

**Comment by noomorph** [comment]
@christianbach, after a dozen of superficial checks - so far so good. We've merged it into our fork, on the next week we'll know better as it will get more collateral QA time.

**Comment by sunnylqm** [comment]
Maybe we need an option to enable purecomponent or not? For example sometimes we need a list that is long but rarely change, purecomponent actually would cause more overhead.

**Comment by axlider** [comment]
Thanks for the great performance optimization. It will definitely help a lot!
But now we need a way to trigger an update on CellRenderer.
Consider this use case (broken use case in my app):
- User clicks on an icon in a row 
- row should be re-renderer e.g because icon should be swapped to reflect selection/deselection.
Now this row won't be re re-rendered anymore, because the item did not change.
Users will be forced to wrap each "item" in "renderItem" in a component and call "this.forceUpdate()" in the component.
At least this should be mentioned in documentation.
Or would it be possible to pass a function in "renderItem", so user can decide whether or not to re-render the item?

**Comment by noomorph** [comment]
@axlider, Alexander, thanks for the feedback, but could you illustrate your example with a screenshot and a JSX code snippet? I'm not sure if I comprehend how your renderItem looks like and what component is inside.

**Comment by sahrens** [comment]
@axlider: you should either keep the selection state in the items, store it in the state of the components that are rendered in renderItem, or use something like Relay or redux to flow the changes through props. No need to add complexity to FlatList to support that usecase I don’t think.

**Comment by axlider** [comment]
@noomorph 
I removed all the noise and this is the essence of my code in 'renderItem'.
'isSelected' refers to a selector in my redux reducer.
Before I applied the changes from pull request this JSX got re-rendered. 

```
...
//In the middle of renderItem...
<TouchableOpacity onPress={() => this.handleItemSelected(item)}>
     {!this.isSelected(item) && <View><EmptyCheckbox/></View>}
     {this.isSelected(item) && <View><CheckedCheckbox/></View>}
</TouchableOpacity>
...
```
Now, my workaround is to extract it into a component and call this.forceUpdate() to get it re-rendered.
```
![2019-02-21 20 53 48](https://user-images.githubusercontent.com/25104854/53197865-67807700-361b-11e9-9e58-2d263dc7e43f.png)

**Comment by noomorph** [comment]
@axlider, if you extract that into a component, why do you need forceUpdate if you can just use `setState`?

**Comment by axlider** [comment]
@noomorph  there are different solutions to force the items to re-render and no problem with that. I just wanted to mention that.
BTW i am very happy with this performance improvements, even if it might break existing apps.

**Comment by reubenab** [comment]
Will this be available in 0.59?

**Comment by christianbach** [comment]
@reubenab don’t think so. FB needs to do some kind of internal change first. The PR is reviewed and ready to go. Just waiting on FB.

**Comment by noomorph** [comment]
@christianbach, I have some news from the battlefield — PR broke several places in our React Native app. Below is a rough illustration of how shortcoming in VirtualizedList turned into a feature many of devs (mis)use:

```jsx
<FlatList data={[1, 2, 3]} renderItem={renderItem} />

const externalDataProvider = {
  getItemById(id) {
    /* ... some logic ... */
    return { id, name: /*...*/, };
  }
};

function renderItem({ item }) {
  return <Text>{externalDataProvider.getItemById(item).name</Text>
}
```

Of course, since items in `data` do not change between app state changes and `renderItem` stays the same, with this PR merged, the `FlatList` does not rerender upon implicit state changes (while previously it understandably did).

As the existing code (that was broken) relied on redundant rerenders, we had to revert your pull request from our fork, primarily because we couldn't assess promptly and precisely what the affected area is.

All the above sounds irrational — don't use bugfix because we rely on the bug — but, unfortunately, here we have to compromise between peace of mind and slight performance improvement.

Speaking for myself, I suggested to add `extraData={{}}` everywhere before the repeated merge and then remove those hacks gradually as lists get tested with and without them. That's still under consideration, so I don't know whether this idea will work out.

**Comment by axlider** [comment]
But if you apply the workarounds we talked about, it will work just fine, even without using extraData.
That's why I extracted "renderItem" into a separate component.
The Items will then re-render either by changed state or props.
It would be sad if this PR would be blocked. People will have to change their code anyway.

**Comment by christianbach** [comment]
@noomorph This made me laugh :) I totally agree that peace of mind is better than a tiny performance improvement.

There seems however to still be a confusion around *when* and *why* the virtualized list should be render/rerender I wonder if there needs to be more clarifying in the docs?

**Comment by noomorph** [comment]
@christianbach, I don't really think you should add more warnings to documentation with this pull request. See the official docs, they already cover this caveat (despite that the caveat was not effective for some time until now):

screen shot 2019-03-07 at 1 57 03 pm

**Comment by sahrens** [comment]
I agree we shouldn't pollute the docs further, but we should definitely call out the changes clearly e.g. in release notes. @hramos: can you also make sure to post internally in case we miss any internal breakage? It should be pretty easy for people to add `extraData={{}}` as a quick workaround.

**Comment by christianbach** [comment]
@sahrens any update on how the `internal diff need some FB love` is coming along?

**Comment by AngefloMusic** [comment]
Any updates on when this will be fixed?

**Comment by noomorph** [comment]
@christianbach, it seems like you have a merge conflict there in `Libraries/Lists/VirtualizedList.js`. I wonder maybe you should update yours?

**Comment by derekstavis** [comment]
@christianbach There's a conflict once again. Can you fix it? Also, as soon as the conflict is resolved, can we merge this one or are we missing anything here?

**Comment by christianbach** [comment]
@derekstavis sure, will try to find some time this week.

**Comment by cpojer** [comment]
The reason we haven't landed this change yet is mainly because nobody at FB has volunteered to ensure that this won't break various surfaces. We have more than 1000 screens written using React Native and this change can potentially break any number of them by not updating things properly and we have no way of automatically detecting it :(

**Comment by sunnylqm** [comment]
How about make this optimization an option but not enabled by default.

**Comment by cpojer** [comment]
If we don't change the default behavior this will be much easier to land, yes!

**Comment by azizhk** [comment]
I tried my hand at not making this the default behavior and accepting an additional prop `flattenParentProps`, but I couldn't figure out how to retain the current `onLayout` behavior.

Changes: https://github.com/christianbach/react-native/compare/virtuallist-optimize-cellrenderer...azizhk:virtuallist-optimize-cellrenderer?expand=1

**Comment by brunolemos** [comment]
>  @cpojer: We have more than 1000 screens written using React Native and this change can potentially break any number of them by not updating things properly

Can't anyone add `extraData={{}}` on all `FlatList`s like Spencer suggested? Should garantee the lists will update and allow merging this pr.

**Comment by cihati** [comment]
@christianbach just wanted to say that I admire your tenacity to get this PR merged (after close to a year!), and I am really looking forward to it! Many thanks for putting in the effort.

**Comment by azizhk** [comment]
I was able to contain most of the changes to be opt-in and not be the default behavior.
The diff with @christianbach 's branch is [here](https://github.com/christianbach/react-native/compare/virtuallist-optimize-cellrenderer...azizhk:virtuallist-optimize-cellrenderer?expand=1) and with facebook's master branch is [here](https://github.com/facebook/react-native/compare/master...azizhk:virtuallist-optimize-cellrenderer?expand=1).

Though this has a bit of [hacky JS context trickery](https://github.com/azizhk/react-native/blob/6fe0e12945f19b72437c4f33027d329d047014f9/Libraries/Lists/VirtualizedList.js#L638-L641
) which might make it brittle. If acceptable, would like some guidance on how to write some good tests so that this does not break easily.

```js
let virtualizedList = this;
this._onCellLayout = function (e) {
  virtualizedList._onCellLayoutGen(e, this.cellKey, this.index)
}
```

**Comment by deorukhkarchaitanya** [comment]
This PR is beautiful. I hope this gets merged. 

CellRenderer re-renders is one of the massive bottlenecks for our performance where we have a VirtualizedList with over 4000 rows.

Thank you everyone who is involved in this PR.  Hope I can help here too.

**Comment by sunnylqm** [comment]
@azizhk If that is too diffcult, how about something like this:
export default (props) => {
   return props.experimentalOptimazation ?  <ExperimentalFlatList {...props} /> : <FlatList {...props} />
}

**Comment by azizhk** [comment]
@sunnylqm Thanks. We also have changes in ScrollViewStickyHeader.js so for that we would need to make changes in a lot of Components (like ScrollView). That gave me a different idea, wrapping inside an isExperimentalOptimization Context Provider. Something like:

```
<VirtualListExperimentalOptimization.Provider>
  <VirtualList {...props} />
</VirtualListExperimentalOptimization.Provider>
```

This can be passed down to ScrollViewStickyHeader. Let me whip up something over the weekend which agrees to the "not changing the default behavior" requirement.

**Comment by brunolemos** [comment]
> @cpojer: If we don't change the default behavior this will be much easier to land, yes!

I have an idea to get this merged:

```diff
-const { extraData } = props
+const { extraData = {} } = props
```

By changing the default value of `extraData` to a new object reference, we keep the current FlatList behavior: update on every render. 

If the user wants to enable the optimization they just need to pass `extraData={null}` or some memoized value.

Ideally this would not be necessary, but since internal facebook code is blocking the merge of this pull request this may be a good middle term solution.

**Comment by azizhk** [comment]
I was able to retain the default behavior for onLayout as well: https://github.com/facebook/react-native/compare/master...azizhk:exp_context_virtuallist

@cpojer Let me know if looks good and I'll create a PR and add tests.

**Comment by christianbach** [comment]
Perhaps it's time to close this PR?

**Comment by cpojer** [comment]
Alright, I'll close it but I still think it is valueable to make this change in a backwards-compatible way – I'm fine with introducing new props with new names that we can migrate people to.

**Comment by sunnylqm** [comment]
@cpojer There is a change in a backwards-compatible way, please review https://github.com/facebook/react-native/compare/master...azizhk:exp_context_virtuallist

**Comment by azizhk** [comment]
@cpojer @christianbach I've created my PR: https://github.com/facebook/react-native/pull/27115. Hoping we can continue the discussion there. Please close it, if it does not seem viable.

**Pull request closed without merging**